File: src/packages/tauv_common/src/state_estimation/ekf.py
from math import sin, cos, tan, pi
from enum import IntEnum
from typing import List, Optional

import rospy
from geometry_msgs.msg import Vector3
import numpy as np


class StateIndex(IntEnum):
    X = 0
    Y = 1
    Z = 2
    VX = 3
    VY = 4
    VZ = 5
    AX = 6
    AY = 7
    AZ = 8
    YAW = 9
    PITCH = 10
    ROLL = 11
    VYAW = 12
    VPITCH = 13
    VROLL = 14


class EKF:
    NUM_FIELDS = 15

    def __init__(self, dvl_offset: np.array, process_covariance: np.array):
        self._dvl_offset: np.array = dvl_offset
        self._process_covariance: np.array = process_covariance

        self._state: np.array = np.zeros(EKF.NUM_FIELDS, float)
        self._covariance: np.array = 1e-9 * np.identity(EKF.NUM_FIELDS, float)

        self._time: Optional[rospy.Time] = None

    def get_state(self, time: rospy.Time) -> (Vector3, Vector3, Vector3, Vector3, Vector3):
        if self._time is None or time <= self._time:
            raise ValueError('get_state for time earlier than last time')

        delta_t: float = time.to_sec() - self._time.to_sec()
        self._time = time

        self._extrapolate_state(delta_t)
        self._extrapolate_covariance(delta_t)

        position = Vector3(self._state[StateIndex.X], self._state[StateIndex.Y], self._state[StateIndex.Z])
        velocity = Vector3(self._state[StateIndex.VX], self._state[StateIndex.VY], self._state[StateIndex.VZ])
        acceleration = Vector3(self._state[StateIndex.AX], self._state[StateIndex.AY], self._state[StateIndex.AZ])
        orientation = Vector3(self._state[StateIndex.ROLL], self._state[StateIndex.PITCH], self._state[StateIndex.YAW])
        angular_velocity = Vector3(self._state[StateIndex.VROLL], self._state[StateIndex.VPITCH], self._state[StateIndex.VYAW])
        return position, velocity, acceleration, orientation, angular_velocity

    def handle_imu_measurement(self, orientation: np.array, angular_velocity: np.array, linear_acceleration: np.array, covariance: np.array, timestamp: rospy.Time):
        if self._time is None:
            self._time = timestamp

        delta_t: float = timestamp.to_sec() - self._time.to_sec()

        self._time = timestamp

        self._extrapolate_state(delta_t)
        self._extrapolate_covariance(delta_t)

        H: np.array = self._get_H([StateIndex.YAW, StateIndex.PITCH, StateIndex.ROLL,
                                   StateIndex.AX, StateIndex.AY, StateIndex.AZ])

        z: np.array = np.concatenate((np.flip(orientation), linear_acceleration))

        y: np.array = z - np.matmul(H, self._state)

        R: np.array = np.diag(covariance)

        S: np.array = np.matmul(H, np.matmul(self._covariance, np.transpose(H))) + R

        K: np.array = np.matmul(self._covariance, np.matmul(np.transpose(H), np.linalg.inv(S)))

        I: np.array = np.identity(EKF.NUM_FIELDS, float)

        self._state = self._state + np.matmul(K, y)
        self._wrap_angles()

        self._covariance = np.matmul(I - np.matmul(K, H), self._covariance)
        self._covariance = np.maximum(np.abs(self._covariance), 1e-9 * np.identity(EKF.NUM_FIELDS, float))

    def handle_dvl_measurement(self, linear_velocity: np.array, covariance: np.array, timestamp: rospy.Time):
        if self._time is None:
            self._time = timestamp

        delta_t: float = timestamp.to_sec() - self._time.to_sec()

        self._time = timestamp

        self._extrapolate_state(delta_t)
        self._extrapolate_covariance(delta_t)

        H: np.array = self._get_H([StateIndex.VX, StateIndex.VY, StateIndex.VZ])

        z: np.array = linear_velocity - self._get_dvl_tangential_velocity()

        y: np.array = z - np.matmul(H, self._state)

        R: np.array = np.diag(covariance)

        S: np.array = np.matmul(H, np.matmul(self._covariance, np.transpose(H))) + R

        K: np.array = np.matmul(self._covariance, np.matmul(np.transpose(H), np.linalg.inv(S)))

        I: np.array = np.identity(EKF.NUM_FIELDS, float)

        self._state = self._state + np.matmul(K, y)
        self._wrap_angles()

        self._covariance = np.matmul(I - np.matmul(K, H), self._covariance)
        self._covariance = np.maximum(np.abs(self._covariance), 1e-9 * np.identity(EKF.NUM_FIELDS, float))

    def handle_depth_measurement(self, depth: float, covariance: float, timestamp: rospy.Time):
        if self._time is None:
            self._time = timestamp

        delta_t: float = timestamp.to_sec() - self._time.to_sec()

        self._time = timestamp

        self._extrapolate_state(delta_t)
        self._extrapolate_covariance(delta_t)

        H: np.array = self._get_H([StateIndex.Z])

        z: np.array = np.array([depth])

        y: np.array = z - np.matmul(H, self._state)

        R: np.array = np.array([covariance])

        S: np.array = np.matmul(H, np.matmul(self._covariance, np.transpose(H))) + R

        K: np.array = np.matmul(self._covariance, np.matmul(np.transpose(H), np.linalg.inv(S)))

        I: np.array = np.identity(EKF.NUM_FIELDS, float)

        self._state = self._state + np.matmul(K, y)
        self._wrap_angles()

        self._covariance = np.matmul(I - np.matmul(K, H), self._covariance)
        self._covariance = np.maximum(np.abs(self._covariance), 1e-9 * np.identity(EKF.NUM_FIELDS, float))


    def _extrapolate_state(self, dt: float):
        F: np.array = self._get_F(dt)

        self._state = np.matmul(F, self._state)
        self._wrap_angles()

    def _wrap_angles(self):
        self._state[StateIndex.YAW] = (self._state[StateIndex.YAW] + pi) % (2 * pi) - pi
        self._state[StateIndex.PITCH] = (self._state[StateIndex.PITCH] + pi) % (2 * pi) - pi
        self._state[StateIndex.ROLL] = (self._state[StateIndex.ROLL] + pi) % (2 * pi) - pi

    def _extrapolate_covariance(self, dt: float):
        J: np.array = self._get_J(dt)

        Q: np.array = self._process_covariance

        self._covariance = np.matmul(J, np.matmul(self._covariance, np.transpose(J))) + Q

    def _get_dvl_tangential_velocity(self) -> np.array:
        cp: float = cos(self._state[StateIndex.PITCH])
        sp: float = sin(self._state[StateIndex.PITCH])
        cr: float = cos(self._state[StateIndex.ROLL])
        sr: float = sin(self._state[StateIndex.ROLL])

        vyaw: float = self._state[StateIndex.VYAW]
        vpitch: float = self._state[StateIndex.VPITCH]
        vroll: float = self._state[StateIndex.VROLL]

        w: np.array = np.array([-sp * vyaw + vroll,
                                cp * sr * vyaw + cr * vpitch,
                                cp * cr * vyaw - sr * vpitch])

        return np.cross(w, self._dvl_offset)

    def _get_H(self, fields: List[int]) -> np.array:
        H: np.array = np.zeros((len(fields), EKF.NUM_FIELDS), float)

        for i, f in enumerate(fields):
            H[i, f] = 1

        return H

    def _get_F(self, dt: float) -> np.array:
        cy: float = cos(self._state[StateIndex.YAW])
        sy: float = sin(self._state[StateIndex.YAW])
        cp: float = cos(self._state[StateIndex.PITCH])
        sp: float = sin(self._state[StateIndex.PITCH])
        tp: float = tan(self._state[StateIndex.PITCH])
        cr: float = cos(self._state[StateIndex.ROLL])
        sr: float = sin(self._state[StateIndex.ROLL])

        F: np.array = np.identity(EKF.NUM_FIELDS, float)
        F[StateIndex.X, StateIndex.VX] = (cp * cy) * dt
        F[StateIndex.X, StateIndex.VY] = (cy * sp * sr - cr * sy) * dt
        F[StateIndex.X, StateIndex.VZ] = (cr * cy * sp + sr * sy) * dt
        F[StateIndex.X, StateIndex.AX] = 0.5 * F[StateIndex.X, StateIndex.VX] * dt
        F[StateIndex.X, StateIndex.AY] = 0.5 * F[StateIndex.X, StateIndex.VY] * dt
        F[StateIndex.X, StateIndex.AZ] = 0.5 * F[StateIndex.X, StateIndex.VZ] * dt
        F[StateIndex.Y, StateIndex.VX] = (cp * sy) * dt
        F[StateIndex.Y, StateIndex.VY] = (cp * cy + sp * sr * sy) * dt
        F[StateIndex.Y, StateIndex.VZ] = (-cy * sr + cr * sp * sy) * dt
        F[StateIndex.Y, StateIndex.AX] = 0.5 * F[StateIndex.Y, StateIndex.VX] * dt
        F[StateIndex.Y, StateIndex.AY] = 0.5 * F[StateIndex.Y, StateIndex.VY] * dt
        F[StateIndex.Y, StateIndex.AZ] = 0.5 * F[StateIndex.Y, StateIndex.VZ] * dt
        F[StateIndex.Z, StateIndex.VX] = (-sp) * dt
        F[StateIndex.Z, StateIndex.VY] = (cp * sr) * dt
        F[StateIndex.Z, StateIndex.VZ] = (cp * cr) * dt
        F[StateIndex.Z, StateIndex.AX] = 0.5 * F[StateIndex.Z, StateIndex.VX] * dt
        F[StateIndex.Z, StateIndex.AY] = 0.5 * F[StateIndex.Z, StateIndex.VY] * dt
        F[StateIndex.Z, StateIndex.AZ] = 0.5 * F[StateIndex.Z, StateIndex.VZ] * dt
        F[StateIndex.VX, StateIndex.AX] = dt
        F[StateIndex.VY, StateIndex.AY] = dt
        F[StateIndex.VZ, StateIndex.AZ] = dt
        F[StateIndex.YAW, StateIndex.VYAW] = (cr / cp) * dt
        F[StateIndex.YAW, StateIndex.VPITCH] = (sr / cp) * dt
        F[StateIndex.PITCH, StateIndex.VYAW] = (-sr) * dt
        F[StateIndex.PITCH, StateIndex.VPITCH] = (cr) * dt
        F[StateIndex.ROLL, StateIndex.VYAW] = (cr * tp) * dt
        F[StateIndex.ROLL, StateIndex.VPITCH] = (sr * tp) * dt
        F[StateIndex.ROLL, StateIndex.VROLL] = dt

        return F

    def _get_J(self, dt: float) -> np.array:
        J: np.array = self._get_F(dt)

        cy: float = cos(self._state[StateIndex.YAW])
        sy: float = sin(self._state[StateIndex.YAW])
        cp: float = cos(self._state[StateIndex.PITCH])
        sp: float = sin(self._state[StateIndex.PITCH])
        tp: float = tan(self._state[StateIndex.PITCH])
        cr: float = cos(self._state[StateIndex.ROLL])
        sr: float = sin(self._state[StateIndex.ROLL])

        J[StateIndex.X, StateIndex.YAW] = self._get_partial(-cp * sy,
                                                            -cy * cr - sp * sy * sr,
                                                            cy * sr - sp * sy * cr,
                                                            dt)
        J[StateIndex.X, StateIndex.PITCH] = self._get_partial(-sp * cy,
                                                              cp * cy * sr,
                                                              cp * cy * cr,
                                                              dt)
        J[StateIndex.X, StateIndex.ROLL] = self._get_partial(0,
                                                             sp * cy * cr + sy * sr,
                                                             sy * cr - sp * cy * sr,
                                                             dt)
        J[StateIndex.Y, StateIndex.YAW] = self._get_partial(cp * cy,
                                                            sp * cy * sr - sy * cr,
                                                            sp * cy * sr + sy * sr,
                                                            dt)
        J[StateIndex.Y, StateIndex.PITCH] = self._get_partial(-sp * sy,
                                                              cp * sy * sr,
                                                              cp * sy * cr,
                                                              dt)
        J[StateIndex.Y, StateIndex.ROLL] = self._get_partial(0,
                                                             sp * sy * cr - cy * sr,
                                                             -cy * cr - sp * sy * sr,
                                                             dt)
        J[StateIndex.Z, StateIndex.PITCH] = self._get_partial(-cp,
                                                              -sp * sr,
                                                              -sp * cr,
                                                              dt)
        J[StateIndex.Z, StateIndex.ROLL] = self._get_partial(0,
                                                             cp * sr,
                                                             -cp * sr,
                                                             dt)

        vyaw: float = self._state[StateIndex.VYAW]
        vpitch: float = self._state[StateIndex.VPITCH]

        J[StateIndex.YAW, StateIndex.PITCH] = (tp * sr / cp) * vpitch * dt + (tp * cr / cp) * vyaw * dt
        J[StateIndex.YAW, StateIndex.ROLL] = (cr / cp) * vpitch * dt + (-sr / cp) * vyaw * dt
        J[StateIndex.PITCH, StateIndex.ROLL] = (-sr) * vpitch * dt + (-cr) * vyaw * dt
        J[StateIndex.ROLL, StateIndex.PITCH] = (sr / (cp * cp)) * vpitch * dt + (cr / (cp * cp)) * vyaw * dt
        J[StateIndex.ROLL, StateIndex.ROLL] = 1 + (tp * cr) * vpitch * dt + (-tp * sr) * vyaw * dt

        return J

    def _get_partial(self, xc: float, yc: float, zc: float, dt: float) -> float:
        vx: float = self._state[StateIndex.VX]
        vy: float = self._state[StateIndex.VY]
        vz: float = self._state[StateIndex.VZ]
        ax: float = self._state[StateIndex.AX]
        ay: float = self._state[StateIndex.AY]
        az: float = self._state[StateIndex.AZ]

        return (xc * vx + yc * vy + zc * vz) * dt + \
               (xc * ax + yc * ay + zc * az) * 0.5 * dt * dt

File: src/packages/tauv_common/src/state_estimation_cpp/ekf.cpp
#include "ekf.h"
#include <ros/ros.h>
#include <eigen3/Eigen/Dense>

Ekf::Ekf()
{
  this->cov = 1e-9 * Eigen::Matrix<double, 15, 15>::Identity();
  this->state = Eigen::Matrix<double, 15, 1>::Zero();
  this->reference_yaw = 0;
}

void Ekf::get_state(Eigen::Matrix<double, 15, 1> &state)
{
  state = this->state;
}

void Ekf::set_state(Eigen::Matrix<double, 15, 1> &state)
{
  this->state = state;
}

void Ekf::get_cov(Eigen::Matrix<double, 15, 15> &cov)
{
  cov = this->cov;
}

void Ekf::set_cov(Eigen::Matrix<double, 15, 15> &cov)
{
  this->cov = cov;
}

void Ekf::set_time(double time)
{
  this->time = time;
}

void Ekf::get_state_fields(double time, Eigen::Vector3d &position, Eigen::Vector3d &velocity, Eigen::Vector3d &acceleration, Eigen::Vector3d &orientation, Eigen::Vector3d &angular_velocity)
{
  using S = StateIndex;

  double dt = time - this->time;

  Eigen::Matrix<double, 15, 1> state = Eigen::Matrix<double, 15, 1>::Zero();
  this->extrapolate_state(dt, this->state, state);

  Eigen::Vector3d corrected_orientation { state(S::ROLL), state(S::PITCH), state(S::YAW) - this->reference_yaw };
  Eigen::Vector3d corrected_position {  cos(-this->reference_yaw) * state(S::X) - sin(-this->reference_yaw) * state(S::Y), sin(-this->reference_yaw) * state(S::X) + cos(-this->reference_yaw) * state(S::Y), state(S::Z) };
  this->wrap_angles(corrected_orientation);

  position << corrected_position.x(), corrected_position.y(), corrected_position.z();
  velocity << state(S::VX), state(S::VY), state(S::VZ);
  acceleration << state(S::AX), state(S::AY), state(S::AZ);
  orientation << corrected_orientation.x(), corrected_orientation.y(), corrected_orientation.z();
  angular_velocity << state(S::VROLL), state(S::VPITCH), state(S::VYAW);
}

void Ekf::handle_imu_measurement(double time,
    const Eigen::Vector3d &orientation,
    const Eigen::Vector3d &rate_of_turn,
    const Eigen::Vector3d &linear_acceleration,
    const Eigen::Matrix<double, 9, 1> &covariance)
{
  if (!this->initialized) {
    this->time = time;
    this->initialized = true;
  }

  this->predict(time);

  Eigen::VectorXi fields(9);
  fields << StateIndex::ROLL, StateIndex::PITCH, StateIndex::YAW, StateIndex::VROLL, StateIndex::VPITCH, StateIndex::VYAW, StateIndex::AX, StateIndex::AY, StateIndex::AZ;

  Eigen::Matrix<double, Eigen::Dynamic, 15> H = Eigen::Matrix<double, Eigen::Dynamic, 15>::Zero(9, 15);
  this->get_H(fields, H);

  Eigen::VectorXd y(9);
  y << orientation, rate_of_turn, linear_acceleration;
  y -= H * this->state;

  Eigen::VectorXi angle_fields(3);
  angle_fields << 0, 1, 2;
//  this->wrap_angles(angle_fields, y);

  Eigen::VectorXd cov(9);
  cov << covariance;

  this->update(fields, y, cov);
}

void Ekf::handle_dvl_measurement(double time, const Eigen::Vector3d &velocity, const Eigen::Vector3d &covariance)
{
  if (!this->initialized) {
    this->time = time;
    this->initialized = true;
  }

  this->predict(time);

  Eigen::VectorXi fields(3);
  fields << StateIndex::VX, StateIndex::VY, StateIndex::VZ;

  Eigen::Matrix<double, Eigen::Dynamic, 15> H = Eigen::Matrix<double, Eigen::Dynamic, 15>::Zero(3, 15);
  this->get_H(fields, H);

  Eigen::Vector3d tangential_velocity = Eigen::Vector3d::Zero();
  this->get_dvl_tangential_velocity(tangential_velocity);

  Eigen::VectorXd y(3);
  y << velocity;
  y -= tangential_velocity;
  y -= H * this->state;

  Eigen::VectorXd cov(3);
  cov << covariance;

  this->update(fields, y, cov);
}

void Ekf::handle_depth_measurement(double time, double depth, double covariance)
{
  if (!this->initialized) {
    this->time = time;
    this->initialized = true;
  }

  this->predict(time);

  Eigen::VectorXi fields(1);
  fields << StateIndex::Z;

  Eigen::Matrix<double, Eigen::Dynamic, 15> H = Eigen::Matrix<double, Eigen::Dynamic, 15>::Zero(1, 15);
  this->get_H(fields, H);

  Eigen::VectorXd y(1);
  y << depth;
  y -= H * this->state;

  Eigen::VectorXd cov(1);
  cov << covariance;

  this->update(fields, y, cov);
}

void Ekf::predict(double time)
{
  double dt = time - this->time;
  this->time = time;

  Eigen::Matrix<double, 15, 1> new_state = Eigen::Matrix<double, 15, 1>::Zero();
  this->extrapolate_state(dt, this->state, new_state);
  this->state = new_state;

  Eigen::Matrix<double, 15, 15> new_cov = Eigen::Matrix<double, 15, 15>::Zero();
  this->extrapolate_cov(dt, this->cov, new_cov);
  this->cov = new_cov;
} 

void Ekf::update(Eigen::VectorXi &fields, Eigen::VectorXd &inn, Eigen::VectorXd &cov)
{
  Eigen::Matrix<double, Eigen::Dynamic, 15> H = Eigen::Matrix<double, Eigen::Dynamic, 15>::Zero(fields.rows(), 15);
  this->get_H(fields, H);

  Eigen::MatrixXd R = Eigen::DiagonalMatrix<double, Eigen::Dynamic, Eigen::Dynamic>(cov);

  Eigen::MatrixXd S = (H * this->cov) * H.transpose() + R;
  Eigen::MatrixXd K = (this->cov * H.transpose()) * S.inverse();

  Eigen::Matrix<double, 15, 15> I = Eigen::Matrix<double, 15, 15>::Identity();

  this->state += K * inn;

  Eigen::VectorXi angle_fields(3); 
  angle_fields << StateIndex::ROLL, StateIndex::PITCH, StateIndex::YAW;
//  this->wrap_angles(angle_fields, this->state);

  this->cov = (I - K * H) * this->cov * (I - K * H).transpose();
  this->cov += (K * R) * K.transpose();
  this->cov = this->cov.cwiseAbs().cwiseMax(1e-9 * I);
}

void Ekf::extrapolate_state(double dt, Eigen::Matrix<double, 15, 1> &old_state, Eigen::Matrix<double, 15, 1> &new_state)
{
  using S = StateIndex;

  Eigen::Matrix<double, 15, 15> F = Eigen::Matrix<double, 15, 15>::Zero();
  this->get_F(dt, F);

  new_state = F * old_state;

  Eigen::VectorXi fields(3); 
  fields << S::ROLL, S::PITCH, S::YAW;
//  this->wrap_angles(fields, new_state);
}

void Ekf::extrapolate_cov(double dt, Eigen::Matrix<double, 15, 15> &old_cov, Eigen::Matrix<double, 15, 15> &new_cov)
{
  Eigen::Matrix<double, 15, 15> J = Eigen::Matrix<double, 15, 15>::Zero();
  this->get_J(dt, J);

  new_cov = J * (old_cov * old_cov.transpose()) + (dt * this->process_covariance);
}

double wrap(double angle)
{
  // https://stackoverflow.com/a/29871193
  return -M_PI + fmod(2 * M_PI + fmod(angle + M_PI, 2 * M_PI), 2 * M_PI);
}

void Ekf::wrap_angles(Eigen::Vector3d &state)
{
  for (int i = 0; i < 3; i++) {
    state(i) = wrap(state(i));
  }
}

void Ekf::wrap_angles(Eigen::VectorXi &fields, Eigen::Matrix<double, 15, 1> &state)
{
  for (int i = 0; i < fields.rows(); i++) {
    state(fields(i)) = wrap(state(fields(i)));
  }
}

void Ekf::wrap_angles(Eigen::VectorXi &fields, Eigen::VectorXd &state)
{
  for (int i = 0; i < fields.rows(); i++) {
    state(fields(i)) = wrap(state(fields(i)));
  }
}

void Ekf::get_dvl_tangential_velocity(Eigen::Matrix<double, 3, 1> &v)
{
  using S = StateIndex;

  double cp = cos(this->state(S::PITCH));
  double sp = sin(this->state(S::PITCH));
  double cr = cos(this->state(S::ROLL));
  double sr = sin(this->state(S::ROLL));

  double vyaw = this->state(S::VYAW);
  double vpitch = this->state(S::VPITCH);
  double vroll = this->state(S::VROLL);

  Eigen::Vector3d w;
  w << (-sp * vyaw + vroll),
    (cp * sr * vyaw + cr * vpitch),
    (cp * cr * vyaw - sr * vpitch);

  v = w.cross(this->dvl_offset);
}

void Ekf::get_H(Eigen::VectorXi &fields, Eigen::Matrix<double, Eigen::Dynamic, 15> &H)
{
  H = Eigen::Matrix<double, Eigen::Dynamic, 15>(fields.rows(), 15) = Eigen::Matrix<double, Eigen::Dynamic, 15>::Zero(fields.rows(), 15); 

  for (int i = 0; i < fields.rows(); i++) {
    H(i, fields(i, 0)) = 1.0; 
  }
}

void Ekf::get_F(double dt, Eigen::Matrix<double, 15, 15> &F)
{
  using S = StateIndex;
  double cy = cos(this->state(S::YAW));
  double sy = sin(this->state(S::YAW));
  double cp = cos(this->state(S::PITCH));
  double sp = sin(this->state(S::PITCH));
  double tp = tan(this->state(S::PITCH));
  double cr = cos(this->state(S::ROLL));
  double sr = sin(this->state(S::ROLL));

  // State evolution matrix comes from robot_localization math
  // https://github.com/cra-ros-pkg/robot_localization/blob/ea976f9e65eac505f5cc7d3197ae080de4e5c10b/src/ekf.cpp#L259C5-L259C5

  // This propagates the last states forward in time, tracking 
  // Position in the world frame
  // Velocity in the body frame
  // Acceleration in the body frame
  // Orientation (euler angles) in the world frame
  // Angular velocity (axis-angle) in the body frame

  // See the GitHub wiki entry on kinematics math for more info on this

  F = Eigen::Matrix<double, 15, 15>::Identity();
  F(S::X, S::VX) = (cp * cy) * dt;
  F(S::X, S::VY) = (cy * sp * sr - cr * sy) * dt;
  F(S::X, S::VZ) = (cr * cy * sp + sr * sy) * dt;
  F(S::X, S::AX) = 0.5 * F(S::X, S::VX) * dt;
  F(S::X, S::AY) = 0.5 * F(S::X, S::VY) * dt;
  F(S::X, S::AZ) = 0.5 * F(S::X, S::VZ) * dt;
  F(S::Y, S::VX) = (cp * sy) * dt;
  F(S::Y, S::VY) = (cp * cy + sp * sr * sy) * dt;
  F(S::Y, S::VZ) = (-cy * sr + cr * sp * sy) * dt;
  F(S::Y, S::AX) = 0.5 * F(S::Y, S::VX) * dt;
  F(S::Y, S::AY) = 0.5 * F(S::Y, S::VY) * dt;
  F(S::Y, S::AZ) = 0.5 * F(S::Y, S::VZ) * dt;
  F(S::Z, S::VX) = (-sp) * dt;
  F(S::Z, S::VY) = (cp * sr) * dt;
  F(S::Z, S::VZ) = (cp * cr) * dt;
  F(S::Z, S::AX) = 0.5 * F(S::Z, S::VX) * dt;
  F(S::Z, S::AY) = 0.5 * F(S::Z, S::VY) * dt;
  F(S::Z, S::AZ) = 0.5 * F(S::Z, S::VZ) * dt;
  F(S::VX, S::AX) = dt;
  F(S::VY, S::AY) = dt;
  F(S::VZ, S::AZ) = dt; 
  F(S::YAW, S::VYAW) = (cr / cp) * dt;
  F(S::YAW, S::VPITCH) = (sr / cp) * dt;
  F(S::PITCH, S::VYAW) = (-sr) * dt;
  F(S::PITCH, S::VPITCH) = (cr) * dt;
  F(S::ROLL, S::VYAW) = (cr * tp) * dt;
  F(S::ROLL, S::VPITCH) = (sr * tp) * dt;
  F(S::ROLL, S::VROLL) = dt;
}

void Ekf::get_J(double dt, Eigen::Matrix<double, 15, 15> &J)
{
  using S = StateIndex;

  this->get_F(dt, J);

  double cy = cos(this->state(S::YAW));
  double sy = sin(this->state(S::YAW));
  double cp = cos(this->state(S::PITCH));
  double sp = sin(this->state(S::PITCH));
  double tp = tan(this->state(S::PITCH));
  double cr = cos(this->state(S::ROLL));
  double sr = sin(this->state(S::ROLL));

  J(S::X, S::YAW) = this->get_partial(-cp * sy,
      -cy * cr - sp * sy * sr,
      cy * sr - sp * sy * cr,
      dt);
    J(S::X, S::PITCH) = this->get_partial(-sp * cy,
        cp * cy * sr,
        cp * cy * cr,
        dt);
    J(S::X, S::ROLL) = this->get_partial(0,
        sp * cy * cr + sy * sr,
        sy * cr - sp * cy * sr,
        dt);
    J(S::Y, S::YAW) = this->get_partial(cp * cy,
        sp * cy * sr - sy * cr,
        sp * cy * sr + sy * sr,
        dt);
    J(S::Y, S::PITCH) = this->get_partial(-sp * sy,
        cp * sy * sr,
        cp * sy * cr,
        dt);
    J(S::Y, S::ROLL) = this->get_partial(0,
        sp * sy * cr - cy * sr,
        -cy * cr - sp * sy * sr,
        dt);
    J(S::Z, S::PITCH) = this->get_partial(-cp,
        -sp * sr,
        -sp * cr,
        dt);
    J(S::Z, S::ROLL) = this->get_partial(0,
        cp * sr,
        -cp * sr,
        dt);

  double vyaw = this->state(S::VYAW);
  double vpitch = this->state(S::VPITCH);

  J(S::YAW, S::PITCH) = (tp * sr / cp) * vpitch * dt + (tp * cr / cp) * vyaw * dt;
  J(S::YAW, S::ROLL) = (cr / cp) * vpitch * dt + (-sr / cp) * vyaw * dt;
  J(S::PITCH, S::ROLL) = (-sr) * vpitch * dt + (-cr) * vyaw * dt;
  J(S::ROLL, S::PITCH) = (sr / (cp * cp)) * vpitch * dt + (cr / (cp * cp)) * vyaw * dt;
  J(S::ROLL, S::ROLL) = 1 + (tp * cr) * vpitch * dt + (-tp * sr) * vyaw * dt;
}

double Ekf::get_partial(double xc, double yc, double zc, double dt)
{
  using S = StateIndex;

  double vx = this->state(S::VX);
  double vy = this->state(S::VY);
  double vz = this->state(S::VZ);
  double ax = this->state(S::AX);
  double ay = this->state(S::AY);
  double az = this->state(S::AZ);

  return (xc * vx + yc * vy + zc * vz) * dt + (xc * ax + yc * ay + zc * az) * 0.5 * dt * dt;
}

void Ekf::set_reference_yaw(double reference_yaw)
{
    this->reference_yaw = reference_yaw;
}

Eigen::Quaterniond rpy_to_quat(const Eigen::Vector3d &rpy)
{
    Eigen::AngleAxisd roll_angle(-rpy.x(), Eigen::Vector3d::UnitX());
    Eigen::AngleAxisd pitch_angle(-rpy.y(), Eigen::Vector3d::UnitY());
    Eigen::AngleAxisd yaw_angle(-rpy.z(), Eigen::Vector3d::UnitZ());

    Eigen::Quaterniond q = roll_angle * pitch_angle * yaw_angle;

    return q;
}